astropy/time/core.py

```diff
@@ -1810,50 +1871,96 @@ def sidereal_time(self, kind, longitude=None, model=None):
 
         if model is None:
             model = sorted(available_models.keys())[-1]
-        else:
-            if model.upper() not in available_models:
-                raise ValueError(
-                    'Model {} not implemented for {} sidereal time; '
-                    'available models are {}'
-                    .format(model, kind, sorted(available_models.keys())))
+        elif model.upper() not in available_models:
+            raise ValueError(
+                'Model {} not implemented for {} sidereal time; '
+                'available models are {}'
+                .format(model, kind, sorted(available_models.keys())))
 
-        if longitude is None:
-            if self.location is None:
-                raise ValueError('No longitude is given but the location for '
-                                 'the Time object is not set.')
-            longitude = self.location.lon
-        elif longitude == 'greenwich':
-            longitude = Longitude(0., u.degree,
-                                  wrap_angle=180. * u.degree)
-        else:
-            # sanity check on input
-            longitude = Longitude(longitude, u.degree,
-                                  wrap_angle=180. * u.degree)
+        model_kwargs = available_models[model.upper()]
+
+        if isinstance(longitude, str) and longitude in ('tio', 'greenwich'):
+            longitude = 0
+            model_kwargs = model_kwargs.copy()
+            model_kwargs['include_tio'] = False
 
-        gst = self._erfa_sidereal_time(available_models[model.upper()])
-        return Longitude(gst + longitude, u.hourangle)
+        return self._st_or_era(longitude=longitude, **model_kwargs)
 
     if isinstance(sidereal_time.__doc__, str):
         sidereal_time.__doc__ = sidereal_time.__doc__.format(
             'apparent', sorted(SIDEREAL_TIME_MODELS['apparent'].keys()),
             'mean', sorted(SIDEREAL_TIME_MODELS['mean'].keys()))
 
-    def _erfa_sidereal_time(self, model):
-        """Calculate a sidereal time using a IAU precession/nutation model."""
+    def _st_or_era(self, longitude, function, scales, include_tio=True):
```

Comment:
Just a nit, but there is no good reason to rename this private method with such a cryptic name. For instance `_sid_time_or_earth_rot_ang` would get the meaning across.

Reply:
Sure.